### ISSUE TITLE
Graphical apps shutdown after sleep mode

### ISSUE THREAD

**Issue by vincowl** [chromebrew/chromebrew#3359]
## Description
Graphical apps are automatically shut down after sleep mode. This occurs at least with Code and Gemacs. It happened after upgrading gtk3 to 3.24.9 and xdg_utils to 1.1.3_1. This happened before this only with gemacs but it let me think it was linkled to this long lasting bug (https://bugzilla.gnome.org/show_bug.cgi?id=85715 and https://gitlab.gnome.org/GNOME/gtk/issues/221)

Output:
```shell
The XKEYBOARD keymap compiler (xkbcomp) reports:
> Error:            Key <CAPS> added to map for multiple modifiers
>                   Using Mod3, ignoring Lock.
Errors from xkbcomp are not fatal to the X server
(EE) failed to read Wayland events: Broken pipe
```

## Versions and system information
chromebrew: `crew --version`
> 1.2.4

ChromeOS: `cat /etc/lsb-release`
> CHROMEOS_ARC_ANDROID_SDK_VERSION=25
> CHROMEOS_ARC_VERSION=5669872
> CHROMEOS_AUSERVER=https://tools.google.com/service/update2
> CHROMEOS_BOARD_APPID={B00DD0BC-D2C9-BAB4-E66C-81AE3F5A7CED}
> CHROMEOS_CANARY_APPID={90F229CE-83E2-4FAF-8479-E368A34938B1}
> CHROMEOS_DEVSERVER=
> CHROMEOS_RELEASE_APPID={B00DD0BC-D2C9-BAB4-E66C-81AE3F5A7CED}
> CHROMEOS_RELEASE_BOARD=cave-signed-mpkeys
> CHROMEOS_RELEASE_BRANCH_NUMBER=75
> CHROMEOS_RELEASE_BUILDER_PATH=cave-release/R75-12105.75.0
> CHROMEOS_RELEASE_BUILD_NUMBER=12105
> CHROMEOS_RELEASE_BUILD_TYPE=Official Build
> CHROMEOS_RELEASE_CHROME_MILESTONE=75
> CHROMEOS_RELEASE_DESCRIPTION=12105.75.0 (Official Build) stable-channel cave 
> CHROMEOS_RELEASE_KEYSET=mp
> CHROMEOS_RELEASE_NAME=Chrome OS
> CHROMEOS_RELEASE_PATCH_NUMBER=0
> CHROMEOS_RELEASE_TRACK=stable-channel
> CHROMEOS_RELEASE_VERSION=12105.75.0
> DEVICETYPE=CHROMEBOOK
> GOOGLE_RELEASE=12105.75.0

Model: Asus C302CA

Architecture: `uname -m`
> x86_64

**Comment by skiman6010** [comment]
This is how they have always worked. My personal theory is that it is linked to our older sommelier version. I don't know if anyone has time to investigate this, but the wayland message printed out is true, the pipe is broken when you enter sleep mode.

**Comment by JL2210** [comment]
The newer Sommelier version requires patched Linux headers.

**Comment by vincowl** [comment]
@JL2210 How do we handle this ? Do these patches come as sommelier's dependencies or do we have to handle this manually ? Maybe it is simply not already available. Could you be a bit clearer please ?

**Comment by JL2210** [comment]
We don't handle it. The Linux sources are kept as-is, which is why the latest version of Sommelier doesn't work. The patches are available from Chromium's source repository (couldn't find them to provide a link) but have to be applied in a specific order.

**Comment by vincowl** [comment]
OK. So we have two choices :
- Apply them at our own risk
- Wait and hope that these patches will be integrated in a future chromeOS version 
What a pity !

**Comment by JL2210** [comment]
No, we have these two choices:

* Apply them at our own risk
* Use the current version of Sommelier, which is tested and works.

Chrome OS doesn't provide the Linux headers, we provide them in Chromebrew.

**Comment by vincowl** [comment]
I must be tired, but sommelier's pipe is broken when you enter sleep mode (using latest version) and you said that we must patch linux headers to avoid this. How can current sommelier's version be considered as "working" then ? Anyway, is the above behavior a bug or not ? If not, is the only solution to fix this to apply given patches ? If it is a bug, why not applying patches to the linux headers only used by chromebrew since they are provided this way ?

**Comment by JL2210** [comment]
The new version of Sommelier requires patches to the Linux headers that we
don't apply. The new version of Sommelier might work.

**Comment by JL2210** [comment]
Pipes are always broken upon entering Sleep Mode. It's not really a thing
we can fix unless we upgrade.

**Comment by vincowl** [comment]
I understand a lot better ! Is this upgraded already planned ?

**Comment by erenoglu** [comment]
One thing I dont get, if existing ChromeOS kernel already have these patches embedded before compiling, it means the running kernel binary has those and we could build new sommelier that will make use of those functions.
Also the kernel in chromeos repositories shall already have that patched version source code. Why can't we build new sommerlier against that patched kernel source (if source is requried for sommelier), and since the kernel already has the function, it would work?

**Comment by JL2210** [comment]
The kernel requires programs to use documented interfaces (the "header
files") to cooperate with it. The documented interfaces provide symbolic
constants for the program to use in order to interact with them, but
currently the documented interfaces aren't available in our kernel headers.
These patches change that.

**Comment by vincowl** [comment]
And why not upgrading our linux headers to include these patches ? Then we could rebuild a new sommelier version and avoid the current broken pipes trouble. Are there foreseen side effects ? I don't know which impact range have these patches on linux headers and what should then been rebuilt.

**Comment by JL2210** [comment]
You're free to create a pull request.

**Comment by vincowl** [comment]
True... But as I'm a developer with few knowledge in sysadmin and kernels area, that was only a question. As already said, I have no idea of what it implies and cannot measure the effort it requires. I have absolutely no problem giving a hand to someone much skilled than I am or receiving a clear answer saying something like : "too much work for now but planned as a (critical ?) request" or "to much work, we don't plan it at the moment". That would even be better than a "get out of that"-like answer.

**Comment by JL2210** [comment]
It's quite literally a 'patch -Np1 -i' command. As these patches are only
adding interfaces, there should be no issue.

**Comment by uberhacker** [comment]
@JL2210: Why don't you create the PR when you get a chance?  This would be a nice fix.

**Comment by JL2210** [comment]
I don't use graphical apps and it would be trivial to do. Along with the
fact that I'm currently busy, someone else could do it, so I don't feel the
need.

**Comment by uberhacker** [comment]
With all the back and forth, you could have just submitted a PR and saved time. :)

**Comment by vincowl** [comment]
@JL2210 Could you give me the name of the package to upgrade please ? If it is as simple as described, I can try to do it. Is this list correct ?
- linuxheaders.rb
- sommelier.rb
- gtk ??

In the self.install method, I should begin with applying patches ? Then how do I check that this works before rebuilding sommelier ?

**Comment by JL2210** [comment]
linuxheaders

GTK has nothing to do with Sommelier, and patches to the Linux headers
aren't supposed to be applied to Sommelier.

**Comment by vincowl** [comment]
@JL2210 Do you think the modification of `self.install` the correct option ?

**Comment by vincowl** [comment]
I tried to apply patches in a `postinstall` step in `linuxheaders.rb` and running `crew install linuxheader`. Several failed due to missing header files (I.e. `qrtr.h`). Despite this, I ran `crew install -s sommelier`. GUI apps are correctly displayed but pipe is always broken  upon entering Sleep Mode. 

Did I miss something ? Any idea ?

**Comment by JL2210** [comment]
How did you apply them? Several require symlinks to work (the Linux headers
have changed from 3.18 to 4.11).

**Comment by JL2210** [comment]
To get the benefit of the Linux header updates, Sommelier needs to be
updated.

**Comment by vincowl** [comment]
I am currently building a branch to have things done in a clean way and to discuss easier. I have probably badly applied patches. What do you mean by "updating" sommelier ? `crew install -s sommelier` is not the right way ?

**Comment by vincowl** [comment]
I am currently working on [this branch in my fork](https://github.com/vincowl/chromebrew/tree/linuxheader-sleep-mode). Shall I create a PR even if it is not working ?

**Comment by vincowl** [comment]
Just wondering what effort it means applying these patches since our headers are from 3.18 and patches relate to 4.11. To me, it means that patch files can not be applied direclty using a `patch -Np1 -i `-like command as previously said. Wouldn't it imply a deep reading and understanding of the headers' modifications from 3.18 to 4.11 ? If so, it's totally out of my field of competence

**Comment by vincowl** [comment]
Someone to give a hand or just hints please ?

**Comment by JL2210** [comment]
@vincowl Give me a little time. I can probably whip up an update.

**Comment by vincowl** [comment]
@JL2210 That would be an amazing thing if this could happen. I'm not in a hurry if this can be successfully handled by someone more skilled than I am.... A big thank for that since you don't need it personally. If there's anything I can do to help, please don't hesitate to ask

**Comment by JL2210** [comment]
@vincowl See #3377.

**Comment by uberhacker** [comment]
@vincowl: Does sommelier crash when your Chromebook goes into sleep mode with the updated linuxheaders package?

**Comment by vincowl** [comment]
@uberhacker I'm on holidays. I'll test it as soon as I'll have hands on my chromebook.

**Comment by uberhacker** [comment]
I can confirm this is still an issue on all architectures even after #3377.

**Comment by JL2210** [comment]
#3377 itself wasn't meant to fix the issue. It just provides the headers
that should allow a Sommelier upgrade.

**Comment by vincowl** [comment]
@JL2210 Shall we just try a `crew install -s sommelier` ?

**Comment by vincowl** [comment]
Previous idea does not work ... as expected...

**Comment by vincowl** [comment]
Upgrading to `59dd453e6efd2c4a0b2b08767c42f47e66a4aaa1` version of sommelier in `https://chromium.googlesource.com/chromiumos/containers/sommelier/` and rebuilding it through an update of `sommelier.rb` doesn't either solve the issue

**Comment by vincowl** [comment]
@JL2210 @uberhacker I am trying to rebuild a new sommelier version using `https://chromium.googlesource.com/chromiumos/platform2/+/refs/heads/master/vm_tools/sommelier/` but this is built using GN files which is a chromium-dedicated tool. Does one of you has any knowledge with this toolchain ? How can I just compile sommelier using the given files ? It seems I must create a `.gn` file and the `BUILD.gn` also has a LOT of dependencies among building and compiling stuff...

**Comment by JL2210** [comment]
crew install gyp

**Comment by vincowl** [comment]
@JL2210 Thanks. But how do I use gyp in that folder since I only have a `BUILD.gn` file to describe how to compile sommelier ?

**Comment by vincowl** [comment]
@JL2210 @uberhacker I am currently working on a way to compile the latest sommellier version. Work can be found [here](https://github.com/vincowl/chromebrew/tree/sommelier-sleep-mode). Shall I create a new issue or a PR even it is not ready to merge ? 

I would appreciate a little help since I have never used `gn`. It seems that I need to git clone an upper level (`platform2`) and compile only sommelier. I found how to create ninja files to build the executable but I have always this error saying : 
```
ERROR at //vm_tools/sommelier/BUILD.gn:16:11: No value named "amd64" in scope "use"
  if (use.amd64) {
          ^----
See //common-mk/gn_root/BUILD.gn:14:5: which caused the file to be included.
    "//${platform_subdir}:all",
    ^-------------------------
sommelier failed to install: `gn gen out --args='pkg_config="pkg-config" libdir="/usr/lib64/" platform_subdir="vm_tools/sommelier" cxx="g++" cc="gcc" ar="ar"'` exited with 1
```
So you have any idea how I could declare `amd64` in the `use` scope ? `common-mk/use.gni` is empty and `amd64` is set by default to false in `common-mk/platform2.py`

**Comment by uberhacker** [comment]
@vincowl: This is just a guess but since your `libdir="/usr/lib64", if anything in the build attempts to write out to this directory, it will fail since it is read-only.  You might try `libdir="/usr/local/lib64"` instead.  Have you tried the `gn` binary that was recently merged into master yet?

**Comment by vincowl** [comment]
@uberhacker You are right. I'll change `libdir` value.
For sure, I am using the gn binary that we merged together yesterday. This the reason that led me to create this package ;-)

**Comment by vincowl** [comment]
So someone has any idea to solve this "amd64 value not found in 'use' scope value" or how to compile sommelier's latest version using gn ? @JL2210, @uberhacker ?

**Comment by JL2210** [comment]
Build all of platform2 with Google's code.

**Comment by vincowl** [comment]
@JL2210 good idea, thanks ! Is this an acceptable solution ? I have no way to build a binary for each cpu architecture and it might be quite long for a single user to install sommelier by compiling the whole platform2 and isolating sommelier's new binary

**Comment by JL2210** [comment]
What I was saying is that it's impossible to isolate Sommelier from the
rest of the build and inefficient to not.

**Comment by vincowl** [comment]
Here is what I tried : get platform2.git, and run `gn  gen out` -> needs to define `platform_subdir` value. I tried setting up `platform_subdir` to `vm_tools/sommelier` which led to previous problem. I also tried to set up `platform_subdir` to `common_mk`, and so on. Can't figure out how to build platform2, there's always something to redefine with a very few infos on how to do it.

**Comment by JL2210** [comment]
Try defining `platform_subdir` to `.`

**Comment by JL2210** [comment]
See https://chromium.googlesource.com/chromiumos/docs/+/631ad338db048ba61c31ff56458e092dd092d604/platform2_primer.md

**Comment by vincowl** [comment]
I am totally stucked. I don' t understand how this `gn` process works and how to make USE flags available since they seem to be properly declared in the `platform2/common-mk/platform2.py`
as described [here](https://chromium.googlesource.com/chromiumos/docs/+/631ad338db048ba61c31ff56458e092dd092d604/platform2_gn.md)
`gn gen out --args='pkg_config="pkg-config" libdir="/usr/lib64/" platform_subdir="vm_tools/sommelier" cxx="g++" cc="gcc" ar="ar"'` produces :
```
ERROR at //vm_tools/sommelier/BUILD.gn:16:11: No value named "amd64" in scope "use"
  if (use.amd64) {
          ^----
See //common-mk/gn_root/BUILD.gn:14:5: which caused the file to be included.
    "//${platform_subdir}:all",
    ^-------------------------
```
while `gn gen out --args='pkg_config="pkg-config" libdir="/usr/lib64/" platform_subdir="vm_tools" cxx="g++" cc="gcc" ar="ar"'` produces 
```
ERROR at //vm_tools/BUILD.gn:9:11: No value named "kvm_host" in scope "use"
  if (use.kvm_host || use.fuzzer) {
          ^-------
See //common-mk/gn_root/BUILD.gn:14:5: which caused the file to be included.
    "//${platform_subdir}:all",
    ^-------------------------
```
and `gn gen out --args='pkg_config="pkg-config" libdir="/usr/lib64/" platform_subdir="." cxx="g++" cc="gcc" ar="ar"'` produces
```
ERROR at //common-mk/gn_root/BUILD.gn:14:5: Unable to load "/usr/local/tmp/crew/null.dir/platform2/BUILD.gn".
    "//${platform_subdir}:all",
    ^-------------------------
```

**Comment by JL2210** [comment]
Just give up. There's certain limitations that you just won't be able to
get around without system support.

**Comment by vincowl** [comment]
@JL2210 Indeed, that's what I was foreseeing. I didn't find where I could submit an issue or get help from the system support. That's too painful to me. You previously said that you didn't use any graphical apps. Do you write any code ? If so, which toolchain are you using ? I suppose it should be vim or emacs if you are not using graphical apps.

**Comment by JL2210** [comment]
@vincowl I write C (and some shell). I often use the GNU toolchain (although I test-compile with other compilers to see if my code works). I use nano, but I should probably get around to learning Vim sometime.

**Comment by satmandu** [comment]
Sorry to necrobump, but did anyone ever find a way to fix the app crashes after waking from sleep? I have the [current version of sommelier](https://github.com/skycocker/chromebrew/pull/4450) working, but I'm definitely getting the same issue with X apps having closed when waking from sleep.

I don't think the crostini sommelier has this issue, but the crostini sommelier has 3x slower graphics acceleration, so I'd prefer to avoid that if possible.

**Comment by vincowl** [comment]
Nope..... I was really hoping that newest sommelier versions could solve this...

**Comment by esprout** [comment]
This issue really should be in the FAQ.  Is there a protocol for adding an entry to the FAQ, or can I just go ahead and edit the page sensibly?

**Comment by satmandu** [comment]
@uberhacker Re: FAQ editing?  I think you can just submit a PR?

@esprout gtk3 apps should no longer crash after sleep as long as they are being loaded as wayland apps, as long as `GDK_BACKEND=wayland` is set, which it is by default.

X programs run through sommelier will still have this issue, but I think that's a sommelier issue we haven't resolved yet.

**Comment by esprout** [comment]
Can any of the web browsers run via wayland?  I have tried firefox and brave, and both couldn't open display without sommelier running.

**Comment by satmandu** [comment]
Try Epiphany.

**Comment by esprout** [comment]
Did a little googling, and verified brave can use wayland via:

brave --enable-features=UseOzonePlatform --ozone-platform=wayland

In theory that will work with any chromium based browser post v87.

**Comment by esprout** [comment]
I just noticed on my Samsung Galaxy Chromebook, sommelier does *not* shut down at suspend, and firefox stays running after resume.  But it does on my HP Chromebook x2.  Any ideas why that is, and how to fix it for the HP x2 so sommelier stays running after suspend?

**Comment by satmandu** [comment]
I think newer kernels don't have that issue.

Nocturne on kernel 4.4 definitely has the sommelier crashing on sleep issue.

**Comment by esprout** [comment]
I see.  The galaxy chromebook is on stable channel, kernel 4.19.166-10987, and the x2 is on beta channel, kernel 4.4.265-20335, so that could be it.  It's a shame these chromebooks don't really get kernel updates.

What's ironic is that the galaxy chromebook randomly reboots when sleeping!  Seems it's a known issue that neither Samsung nor Google has fixed for months now, and who knows if it will ever be fixed.  The random reboots during sleep may only occur in dev mode, but I'd have to take it out of dev mode, and thus wipe everything, to test it.

**Comment by satmandu** [comment]
I'm at least excited to see confirmation that the kernel version COULD be the culprit for the sommelier issue!

**Comment by satmandu** [comment]
This appears not to be an issue after updates from kernel 4.4.
